pin-project-internal/src/pin_project/mod.rs

```diff
@@ -109,21 +112,133 @@ impl Context {
+                }
+
+                #[cfg(not(feature = "RUSTC_IS_NIGHTLY"))]
+                {
+                    Span::call_site()
+                }
+            };
+
+            let struct_ident = Ident::new(&format!("__UnpinStruct{}", orig_ident), make_span());
+            let always_unpin_ident = Ident::new("AlwaysUnpin", make_span());
+
+            // Generate a field in our new struct for every
+            // pinned field in the original type
+            let fields: Vec<_> = self
+                .pinned_fields
+                .iter()
+                .enumerate()
+                .map(|(i, ty)| {
+                    let field_ident = format_ident!("__field{}", i);
+                    quote! {
+                        #field_ident: #ty
+                    }
+                })
+                .collect();
+
+            // We could try to determine the subset of type parameters
+            // and lifetimes that are actually used by the pinned fields
+            // (as opposed to those only used by unpinned fields).
+            // However, this would be tricky and error-prone, since
+            // it's possible for users to create types that would alias
+            // with generic parameters (e.g. 'struct T').
+            //
+            // Instead, we generate a use of every single type parameter
+            // and lifetime used in the original struct. For type parameters,
+            // we generate code like this:
+            //
+            // ```rust
+            // struct AlwaysUnpin<T: ?Sized>(PhantomData<T>) {}
+            // impl<T: ?Sized> Unpin for AlwaysUnpin<T> {}
+            //
+            // ...
+            // _field: AlwaysUnpin<(A, B, C)>
+            // ```
+            //
+            // This ensures that any unused type paramters
+            // don't end up with Unpin bounds
+            let lifetime_fields: Vec<_> = self
+                .generics
+                .lifetimes()
+                .enumerate()
+                .map(|(i, l)| {
+                    let field_ident = format_ident!("__lifetime{}", i);
+                    quote! {
+                        #field_ident: &#l ()
+                    }
+                })
+                .collect();
+
+            let scope_ident = format_ident!("__unpin_scope_{}", orig_ident);
+
+            let full_generics = &self.generics;
+            let mut full_where_clause = where_clause.clone();
+
+            let unpin_clause: WherePredicate = syn::parse_quote! {
+                #struct_ident #ty_generics: ::core::marker::Unpin
+            };
+
+            full_where_clause.predicates.push(unpin_clause);
+
+            let inner_data = quote! {
+
+                struct #always_unpin_ident <T: ?Sized> {
+                    val: ::core::marker::PhantomData<T>
+                }
+
+                impl<T: ?Sized> ::core::marker::Unpin for #always_unpin_ident <T> {}
+
+                pub struct #struct_ident #full_generics #where_clause {
```

Comment:
Could you use struct/enum's visibility instead of `pub`? (add a field for this to Context)

Reply:
Okay, thanks for the explanation.